zcash_primitives/src/merkle_tree/incremental.rs

```diff
@@ -248,4 +252,198 @@ mod tests {
             assert_eq!(read, original);
         }
     }
+
+    #[test]
+    fn test_bridge_roundtrip() {
+        let mut t: BridgeTree<TestNode, 8> = BridgeTree::new(10);
+        let mut to_unwitness = vec![];
+        let mut has_auth_path = vec![];
+        for i in 0usize..100 {
+            assert!(
+                t.append(&TestNode(i.try_into().unwrap())),
+                "Append should succeed."
+            );
+            if i % 5 == 0 {
+                t.checkpoint();
+            }
+            if i % 7 == 0 {
+                t.witness();
+                if i > 0 && i % 2 == 0 {
+                    to_unwitness.push(Position::from(i));
+                } else {
+                    has_auth_path.push(Position::from(i));
+                }
+            }
+            if i % 11 == 0 && !to_unwitness.is_empty() {
+                let pos = to_unwitness.remove(0);
+                t.remove_witness(pos);
```

Comment:
This always removes the witnesses in the order they were added. Should it test removing witnesses out-of-order?

Reply:
Good call. The order will still need to be deterministic, but permuting the order of `to_unwitness` between each iteration will make this test more robust.